File: Data-Structures-and-Algorithms/0020-ValidParentheses/isValid.js
var isValid = function(s) {
    // make a hashmap first for storing the bracket types
    const refmap = {'(': ')', '{': '}', '[': ']'};
    // make the stack
    const stack = [];
    // for each character in the string
    for (var i of s) {
        // if it is an opening bracket, we push it in to the stack for use later on when we pop it
        if (i in refmap) {
            stack.push(i)
        } 
        // and it is a closing one, we pop the stack, and use the returned opening bracket to find the matching closing brakcet in the refmap
        else if (i !== refmap[stack.pop()]) {
            return false
        };
    }
    // if there is an odd number of characters, there will be some stacks lef
    return stack.length === 0;
};

let gs = '()()(){}{}{}[][][]'
let bs = '()()(){}{}{}[][][}'
console.log(isValid(gs))
console.log(isValid(bs))

// O(N) one pass traverse
// O(N) storing stack for the res is just a boolean
// Are the brackets closed properly?

File: Data-Structures-and-Algorithms/0020-ValidParentheses/isValid.py
class Solution(object):
    def isValid(self, s):
        if len(s) < 2:
            return False
        stack = []
        for char in s:
            # if its first character and it is a closing bracket, return false
            if len(stack) == 0:
                if char == ')' or char == '}' or char == ']':
                    return False

            # append opening ones to the stack
            if char == '(' or char == '{' or char == '[':
                stack.append(char)
            
            # if i encounter a closing bracket, i check the last element of the stack
            if char == ')' and stack[-1] == '(':
                stack.pop()
            elif char == ']' and stack[-1] == '[':
                stack.pop()
            elif char == '}' and stack[-1] == '{':
                stack.pop()
            else:
                return False 
        if len(stack) != 0:
            return False
        return True

gs = '()()(){}{}{}[][][]'
bs = '()()(){}{}{}[][][}'
print(Solution().isValid(gs))
print(Solution().isValid(bs))

# O(N) one pass traverse
# O(N) storing stack for the res is just a boolean
# Are the brackets closed properly?